protocol/encoder/encoder.go

```diff
@@ -0,0 +1,82 @@
+package encoder
+
+import (
+	"encoding/binary"
+)
+
+type Encoder struct {
+	bytes  []byte
+	offset int
+}
+
+func NewEncoder(bytes []byte) *Encoder {
```

Comment:
```suggestion
func NewEncoder() *Encoder {
```

Don't think we need `bytes` to be passed in? Could just instantiate an array here